community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/ast/QueryTagger.scala

```
@@ -0,0 +1,230 @@
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v2_3.ast
+
+import org.neo4j.cypher.internal.compiler.v2_3.parser.{CypherParser, ParserMonitor}
+
+import scala.annotation.tailrec
+
+sealed class QueryTag(aName: String) {
+  val name = aName.trim.toLowerCase
+  val token = s":$name"
+
+  override def toString = token
+}
+
+object QueryTags {
+  val all = Set[QueryTag](
+    MatchTag,
+    OptionalMatchTag,
+    RegularMatchTag,
+    ShortestPathTag,
+    NamedPathTag,
+    SingleLengthRelTag,
+    VarLengthRelTag,
+    DirectedRelTag,
+    UnDirectedRelTag,
+    SingleNodePatternTag,
+    RelPatternTag,
+
+    WhereTag,
+    WithTag,
+    ReturnTag,
+    StartTag,
+    UnionTag,
+    UnwindTag,
+    LoadCSVTag,
+    UpdatesTag,
+
+    ComplexExpressionTag,
+    FilteringExpressionTag,
+    LiteralExpressionTag,
+    ParameterExpressionTag,
+    IdentifierExpressionTag
+  )
+
+  private val tagsByName: Map[String, QueryTag] = all.map { tag => tag.name -> tag }.toMap
+
+  def parse(text: String) = {
+    val tokens = tokenize(text.trim)
+    val tags = tokens.map { piece =>
+      tagsByName.getOrElse(piece, throw new IllegalArgumentException(s":$piece is an unknown query tag"))
+    }
+    tags
+  }
+
+  @tailrec
+  private def tokenize(input: String, tags: Set[String] = Set.empty): Set[String] = {
```

Not quite, it will silently skip all non-matches but we want to throw.